### ISSUE TITLE
Docker PiHole v5.2.1 appears to break conditional forwarding

### ISSUE THREAD

**Issue by Paraphraser** [pi-hole/docker-pi-hole#718]
I'm running PiHole on a Raspberry Pi 4 under Docker as part of [SensorsIot/IOTstack](https://github.com/SensorsIot/IOTstack). The `docker-compose.yml` fragment is

```
pihole:
    container_name: pihole
    image: pihole/pihole:latest
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "8089:80/tcp"
    env_file:
      - ./services/pihole/pihole.env
    volumes:
       - ./volumes/pihole/etc-pihole/:/etc/pihole/
       - ./volumes/pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
    dns:
      - 127.0.0.1
      - 1.1.1.1
    # Recommended but not required (DHCP needs NET_ADMIN)
    #   https://github.com/pi-hole/docker-pi-hole#note-on-capabilities
    cap_add:
      - NET_ADMIN
    restart: unless-stopped
```

The above fragment is "as is" from the SensorsIot/IOTstack repo. 

The active variables in my ENV file are:

```
TZ=Australia/Sydney
WEBPASSWORD=******
INTERFACE=eth0
```

The only change I've made to the default ENV file from the SensorsIot/IOTstack repo is the TZ. And, for the record, the RPi **is** connected to Ethernet.

I only use PiHole as a blocker. I do not use it for local DNS or DHCP. I do, however, have a local upstream DNS server authoritative for a local domain. PiHole has conditional forwarding enabled for the local /24 network and local domain suffix, pointing to the local upstream DNS server (bind 9 running on a Mac).

Earlier today I did a `docker-compose pull` which pulled down v5.2.1 and the result was that forwarding of queries to the local upstream DNS stopped working.

I had people yelling at me so I did not have time to dig around or gather evidence. I didn't even glance at the PiHole GUI. I focused on Docker. Recovering needed:

1. Pinning to v5.2; **and**
2. Reloading persistent store from backup.

> Key point: just pinning to 5.2 was not enough so it looks to me that whatever 5.2.1 did to "break" local forwarding was persistent.

This issue is simply a "heads up" that all may not be well with 5.2.1. I hope there's enough info here for someone to be replicate my experience and figure out how to resolve the problem.

This issue may also serve as an anchor for others to add "me too".

For my part, I'm just going to leave PiHole pinned to 5.2 and move on with my life.

**Comment by PromoFaux** [comment]
Thanks for the report. Investigating this now.

**Comment by PromoFaux** [comment]
@Paraphraser how did you have the conditional forwarding set, just via the web interface?

I've a feeling that settings like this have never persisted through docker container versions - but I've not been using it all that long so I cannot be authoritative on that.

One thing to note is that setting it via environment variables persists between containers; see the [`REV_SERVER*`](https://github.com/pi-hole/docker-pi-hole#environment-variables) here

I have just tested the latter method using both the new `REV_SERVER*` variables and the deprecated `CONDITIONAL_FORWARDING*` variables

**Comment by Paraphraser** [comment]
Thanks for replying.

Yes - all configured from the GUI.

> I'm not CLI-averse. I'm an old guy who wrote his first line of code in the 1970s when CLIs were all there was. I CLI-tinker with other parts of IOTstack all the time but, when it comes to PiHole, I just want it to work and block ads. End of. Way back when I first set it up, I just poked and prodded the GUI until it did what I wanted and forever after held my peace, as it were.

For the record, PiHole » Settings » DNS tab:

* Upstream DNS Servers
	- IPv4 for Google (ECS) and Cloudflare (all others disabled).
* Custom upstream DNS Servers
	- all disabled
* Interface listening behavior
	- Listen only on interface eth0
* Advanced DNS settings
	- all disabled
* Conditional forwarding
	- Use Conditional Forwarding checkbox = enabled
	- Local network in CIDR notation: 192.168.132.0/24
	- IP address of your DHCP server (router): 192.168.132.65
	- Local domain name (optional): sub.domain.com <sup>†</sup>

	† name changed in this issue to protect the innocent but that's the basic structure.

It's been that way for more than a year now and Docker-PiHole has been updated several times over that period. If "non-persistence across upgrades" is the explanation then this is the first time it has been apparent to me, so I come back to "something about 5.2.1".

<hr>

I just did an experiment on a configured-alike test RPi4. I let `docker-compose` upgrade to 5.2.1 again then opened the GUI. You are correct. With reference to the settings I listed earlier:

* everything **except** the "Conditional forwarding" grouping persisted across the upgrade.
* all the "Conditional forwarding" controls were reset to defaults (checkbox off, all fields containing placeholder material).

Then I implemented the `REV_SERVER` environment variables and tried again. That repopulated the GUI correctly so I'm assuming everything would work properly were I to repeat this on the "live" RPi4..

I'm quite happy to adopt the `REV_SERVER` environment variables but it would be nice if the GUI still worked.

<hr>

While on this topic, it's only a minor point but it has always seemed to me that the GUI prompt:

> IP address of your DHCP server (router)

and the documentation for `CONDITIONAL_FORWARDING_IP` aka `REV_SERVER_TARGET` talking about "DHCP server" and "router" is a bit misleading.

My reasoning is:

* it's not a **DHCP** server that needs to get this traffic. It's a **DNS** server.
* A router **may** know how to forward DNS traffic directed to the router correctly but it's not guaranteed.
	- In my own case, my router forwards **all** DNS traffic directed to it, to my local upstream DNS server, but that's only because it's how I configured the router;
	- Out of the box, my router would forward DNS traffic directed to it, to my ISP's DNS servers. I presume most off-the-shelf routers would do the same until configured otherwise.

In any event, I see the extra hop going via the router as sub-optimal, so I "go direct" to the local upstream DNS server.

I keep thinking I must be missing some critical point about DHCP in all this. Any thoughts, advice or theories?

Even if I am being obtuse and "DHCP" and "router" are correct in this context, it seems to me that "local DNS server" is equally correct and probably should be mentioned as an option. Yes/no?

**Comment by PromoFaux** [comment]
>it would be nice if the GUI still worked.

Yeah, I hear you here. There have been a few complaints about this, and it's something I keep meaning to look at (as I get more familiar with how the docker container is built) - Usually all my development goes into the core repos.

RE: your other point - I see what you're saying. Keep in mind that Pi-hole, at it's core, was originally designed to be used by people that have next to 0 experience in networking related tools/terminology. Things have organically grown over the years, conditional forwarding was added by a community contributor, and has been tweaked a bit in this release (hence the `REV_SERVER` variables now instead of `CONDITIONAL_FORWARDING*`), but we are the first to admit that there can always be room for improvement. Especially when it comes to ambiguous terminology.

I guess something could be added into the copy above the settings boxes for conditional forwarding. Feel free to suggest something over at https://github.com/pi-hole/adminLTE, or even better make a pull request with suggested changes - we don't bit too hard!

**Comment by TheGITofTeo997** [comment]
Hello guys, i just wanted to 'up' this because i've been also facing this for the last 2 hours without understanding what i was doing wrong. I've tried everything both with the new and the deprecated variables via the web interface and by editing the setupVars.conf, but every time i reboot my pi the conditional forwarding gets unticked and does not work.

**Comment by PromoFaux** [comment]
FWIW, my (working) compose file:

```yml
  pihole:
    container_name: pihole
    hostname: pihole-docker
    depends_on:
      - unbound
    image: pihole/pihole:v5.2.1
    environment:
      - TZ=Europe/London      
      - ServerIP=192.168.1.253
      - DNS1=8.8.8.8
      - DNS2=8.8.4.4
      - REV_SERVER=true
      - REV_SERVER_DOMAIN=lan
      - REV_SERVER_TARGET=192.168.0.1
      - REV_SERVER_CIDR=192.168.0.0/16
      - SKIPGRAVITYONBOOT=1
    volumes:
      - ./pihole/configs/pihole/:/etc/pihole/
      - ./pihole/configs/dnsmasq.d:/etc/dnsmasq.d/
    ports:
      - "53:53/tcp"
      #- "67:67/udp" # Uncomment if you want to use Pi-Hole for DHCP
      - "53:53/udp"
    expose:
       - "80"
    mac_address: d0:ca:ab:cd:ef:fe
    networks:
      home:
        ipv4_address: 192.168.1.253
```

Please also see my notes above about `setupVars.conf` persistence. It's certainly an area for improvement.

**Comment by Paraphraser** [comment]
@TheGITofTeo997 

> editing the setupVars.conf, but every time i reboot my pi the conditional forwarding gets unticked and does not work.

In my case, using environment variables as recommended by @PromoFaux persists across `down` + `up` and, both by inference and confirmed by test, RPi restart too.

**Comment by xBelladonna** [comment]
I can confirm I also have this issue, affecting more than just conditional forwarding. I first noticed this when watchtower pulled a new image and restarted my container, disabling both DNSSEC and conditional forwarding, and preventing me from accessing several local services by hostname.

It appears this is by design in the `start.sh` script which is run whenever the container starts (or restarts). The script first exports all the environment variables designed to set these settings and then calls function `change_setting` from `webpage.sh` (in the base pi-hole repo) which edits `setupVars.conf` to reflect the values of the exported environment variables, regardless of whether they are defined or not. The behaviour I'm referring to starts [here](https://github.com/pi-hole/docker-pi-hole/blob/89972b3ce546606e9e87eb8e7c4d0155e75f9692/start.sh#L57).
Thus, even if `setupVars.conf` had these values set, they will all be unset by the startup script and written to `setupVars.conf` effectively erasing any configuration set using the web interface. 

At first glance, it seems like a simple check before calling `change_setting` to see if these environment variables are undefined (i.e. `[ ! -z ${VARIABLE+x} ]`) would solve the problem but something tells me it would have already been implemented if it was that simple. There is a TODO [here](https://github.com/pi-hole/docker-pi-hole/blob/89972b3ce546606e9e87eb8e7c4d0155e75f9692/start.sh#L45) that would prevent these functions being called if it's not the container's first time starting however this would still cause problems in my case as watchtower routinely re-creates containers when pulling new images, leading to sudden and unexpected resetting of all my config unless I set the environment variables.

I'm currently working around the issue by modifying `start.sh` to include the check I mentioned and mounting it into the container, replacing the existing script. The check I mentioned explicitly looks for undefined variables and returns false if the variable is set to an empty string, so the environment variables can be explicitly set to empty strings and override the existing config and only get ignored if they aren't defined in the first place. If this would be a viable fix then let me know and I can easily PR.
